Make MaybeUninit #[repr(transparent)]

mjbshaw opened the pull request:
Tracking issue: #60405

rust-highfive commented:
r? @dtolnay

(rust_highfive has picked a reviewer for you, use r? to override)

mjbshaw commented:
cc @Centril @RalfJung

Centril commented:
r? @cramertj

cramertj commented:
We discussed this in the lang team meeting today, and the conclusion was that we'd like to document and guarantee that `MaybeUninit<T>` share's `T`'s size, alignment, and ABI when being passed *directly* by-value (not necessarily as a member of another type).

It's unclear exactly what we might want `#[repr(transparent)]` to mean on unions, or if we should even allow it. All existing uses of `#[repr(transparent)}` also grant that e.g. `Option<Wrapper<T>>` has the same representation as `Option<T>`, but this is not true for `#[repr(transparent)]` unions, and cannot be true. With that in mind, we'd like to note that `MaybeUninit<T>` is not guaranteed to be `#[repr(transparent)]`, but only the three above properties (size, align, direct by-value ABI).

Implementation wise, this works out to exactly what this PR does-- sticking `#[repr(transparent)]` on the type, so I think we're all okay with merging this so long as we document it appropriately.

rkruppe commented:
> It's unclear exactly what we might want `#[repr(transparent)]` to mean on unions, or if we should even allow it. All existing uses of `#[repr(transparent)}` also grant that e.g. `Option<Wrapper<T>>` has the same representation as `Option<T>`, but this is not true for `#[repr(transparent)]` unions, and cannot be true. With that in mind, we'd like to note that `MaybeUninit<T>` is not guaranteed to be `#[repr(transparent)]`, but only the three above properties (size, align, direct by-value ABI).

So, uh, this sounds like y'all don't want RFC 2645 after all. Not that I disagree, but I'm surprised to hear this so soon after the RFC was merged, especially since this exact concern was brough up and marked resolved during the FCP. Or do I misunderstand?


> Implementation wise, this works out to exactly what this PR does-- sticking `#[repr(transparent)]` on the type, so I think we're all okay with merging this so long as we document it appropriately.

Note that the `repr(transparent)` attribute is shown in rustdoc, so this is not just an implementation concern. More generally, one argument in favor of the RFC was that having this attribute would be a cleaner way for (programmers, docs, tools, etc.) to learn about the size/align/by-value-ABI transparency vs just stating those guarantees in prose as one-off properties of `MaybeUninit`. If we end up making MaybeUninit special anyway, the rationale for transparent unions is even weaker than it IMO already way.

Centril commented:
> So, uh, this sounds like y'all don't want RFC 2645 after all. Not that I disagree, but I'm surprised to hear this so soon after the RFC was merged, especially since this exact concern was brough up and marked resolved during the FCP. Or do I misunderstand?

I still want the RFC and to affirm the guarantee through an attribute eventually.

> If we end up making MaybeUninit special anyway, the rationale for transparent unions is even weaker than it IMO already way.

Our discussions was mostly around landing this PR before we stabilize the attribute itself, which I do think we should eventually.

cramertj commented:
@rkruppe yeah to be clear, I don't think we're going back on what we said in that RFC, only saying that we don't want this particular addition to be a de-facto stabilization of that RFC's behavior.

rkruppe commented:
Thanks for the clarification. I would like some further clarification on one point: was this decision made aware of the fact that `repr(transparent)` is shown on the type in rustdoc? I, personally, do worry that this makes adding the attribute a de-facto stabilization of transparent unions as a concept, and by extension (since the surprise around `Option` is fundamental and everything else is dead obvious) the semantics currently given to them.

Centril commented:
@rkruppe 

> was this decision made aware of the fact that `repr(transparent)` is shown on the type in rustdoc?

I don't believe this fact was brought up in the discussions.

RalfJung commented:
> Implementation wise, this works out to exactly what this PR does

Notice that this PR is effectively a NOP -- `repr(transparent)` does not actually change how we compute layout, it just adds some extra checks that ensure that we end up in the "happy path".

mjbshaw commented:
I have revised the documentation on `MaybeUninit` that hopefully addresses some of the concerns here. If it is insufficient or requires more work, let me know.

cramertj commented:
Nice!

@bors r+ rollup

bors commented:
:pushpin: Commit 0f9dc6c48e51637a5d54fc791df94d104c1e0aee has been approved by `cramertj`

<!-- @bors r=cramertj 0f9dc6c48e51637a5d54fc791df94d104c1e0aee -->
<!-- homu: {"type":"Approved","sha":"0f9dc6c48e51637a5d54fc791df94d104c1e0aee","approver":"cramertj"} -->

RalfJung commented on src/libcore/mem/maybe_uninit.rs:
```diff
@@ -191,9 +191,23 @@ use crate::mem::ManuallyDrop;
+///
+/// While `MaybeUninit` is `#[repr(transparent)]` (indicating it guarantees the same size,
+/// alignment, and ABI as `T`), this does *not* change any of the previous caveats. `Option<T>` and
+/// `Option<MaybeUninit<T>>` may still have different sizes, and types containing a field of type
```
This is odd. We just said two paragraphs earlier that this is the case, why repeat it here?

I think it makes more sense to change the previous paragraph (starting "However remember that a type *containing* a `MaybeUninit<T>`") to say something like "Despite being `repr(transparent)`, ..."

RalfJung replied:
@mjbshaw can you file a follow-up PR to improve the wording here or should I do that?
